3p/ampanalytics-lib.js

```diff
@@ -0,0 +1,248 @@
+/**
+ * Copyright 2017 The AMP HTML Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS-IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+import './polyfills';
+import {dev} from '../src/log';
+import {initLogConstructor, setReportError} from '../src/log';
+import {reportError} from '../src/error';
+import {IframeMessagingClient} from './iframe-messaging-client';
+import {MessageTypes} from '../src/3p-analytics-common';
+initLogConstructor();
+setReportError(reportError);
+
+/** @private @const {string} */
+const TAG_ = 'ampanalytics-lib';
+
+/**
+ * Receives messages bound for this cross-domain iframe, from all creatives
+ */
+class AmpAnalytics3pMessageRouter {
+  /** @param {!Window} win */
+  constructor(win) {
+
+    /** @private {!Window} */
+    this.win_ = win;
+
+    /** @private {string} */
```

Comment:
Comment done.
Not sure what you mean about the type...a frame can send multiple responses. If the order goes:
1. creative 1 creates xframe
2. creative 2 wants to use xframe
3. xframe renders
...then xframe will send ready messages to both creatives in order to release their queues